### PR TITLE
Prepared vk/commitment types, and individual opening challenges

### PR DESCRIPTION
**Pull request by npwardberkeley**

### CODE REVIEW

File: src/kzg10/data_structures.rs [hunk 1]
```diff
@@ -86,6 +87,39 @@ impl<E: PairingEngine> ToBytes for VerifierKey<E> {
     }
 }
 
+/// `PreparedVerifierKey` is the fully prepared version for checking evaluation proofs for a given commitment.
+/// We omit gamma here for simplicity.
+#[derive(Derivative)]
+#[derivative(Default(bound = ""), Clone(bound = ""), Debug(bound = ""))]
+pub struct PreparedVerifierKey<E: PairingEngine> {
+    /// The generator of G1.
+    pub prepared_g: Vec<E::G1Affine>,
```
Comment (Pratyush) [inline src/kzg10/data_structures.rs:96]: Why not set this to `Vec<E::G1Prepared>`?

File: src/lib.rs [hunk 2]
```diff
@@ -16,13 +16,26 @@ extern crate derivative;
...
+};
+
+#[cfg(feature = "std")]
+use std::{
     collections::{BTreeMap, BTreeSet},
-    iter::FromIterator,
+    rc::Rc,
     string::{String, ToString},
     vec::Vec,
 };
```
Comment (Pratyush) [inline src/lib.rs:41]: This change slipped in incorrectly; we want to just import things from `ark_std`

File: src/lib.rs [hunk 3]
```diff
@@ -74,13 +93,14 @@ pub mod sonic_pc;
 /// The construction is detailed in [[BCMS20]][pcdas].
 ///
 /// [pcdas]: https://eprint.iacr.org/2020/499
+// TODO: add "Prepared" to marlin_pc
```
Comment (Pratyush) [inline src/lib.rs:96]: This TODO is in the wrong place, no?

File: src/lib.rs [hunk 4]
```diff
@@ -74,13 +93,14 @@ pub mod sonic_pc;
...
 pub mod ipa_pc;
 
 /// `QuerySet` is the set of queries that are to be made to a set of labeled polynomials/equations
 /// `p` that have previously been committed to. Each element of a `QuerySet` is a `(label, query)`
 /// pair, where `label` is the label of a polynomial in `p`, and `query` is the field element
 /// that `p[label]` is to be queried at.
-pub type QuerySet<'a, F> = BTreeSet<(String, F)>;
+pub type QuerySet<'a, F> = BTreeSet<(String, (String, F))>;
```
Comment (Pratyush) [inline src/lib.rs:103]: Can you please also update the documentation? What's the extra `String` term for now? Should this be a `BTreeMap` instead?

File: src/lib.rs [hunk 5]
```diff
@@ -96,6 +116,15 @@ pub struct BatchLCProof<F: Field, PC: PolynomialCommitment<F>> {
...
 }
 
+impl<F: Field, PC: PolynomialCommitment<F>> Clone for BatchLCProof<F, PC> {
+    fn clone(&self) -> Self {
+        BatchLCProof {
+            proof: self.proof.clone(),
+            evals: self.evals.clone(),
+        }
+    }
+}
```
Comment (Pratyush) [inline src/lib.rs:126]: Can you derive Clone? Either via the standard macro, or via the `derivative` macro?

File: src/lib.rs [hunk 6]
```diff
@@ -108,9 +137,13 @@ pub trait PolynomialCommitment<F: Field>: Sized {
     /// open the commitment to produce an evaluation proof.
     type CommitterKey: PCCommitterKey;
     /// The verifier key for the scheme; used to check an evaluation proof.
-    type VerifierKey: PCVerifierKey;
+    type VerifierKey: PCVerifierKey + Default;
```
Comment (Pratyush) [inline src/lib.rs:140]: Hmm the default bound on the `VerifierKey` might be incorrect for `ipa_pc::VerifierKey`, because that will just make an empty vector.

File: src/lib.rs [hunk 7]
```diff
@@ -247,14 +282,14 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
-    fn check<'a>(
+    fn check<'a, R: RngCore>(
         vk: &Self::VerifierKey,
         commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
         point: F,
         values: impl IntoIterator<Item = F>,
         proof: &Self::Proof,
         opening_challenge: F,
-        rng: Option<&mut dyn RngCore>,
+        rng: &mut R,
```
Comment (Pratyush) [inline src/lib.rs:292]: This change also seems like a mixup during merge; can you change it to what's on master?

File: src/lib.rs [hunk 8]
```diff
@@ -306,15 +344,7 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
-            result &= Self::check(
-                vk,
-                comms,
-                *query,
-                values,
-                &proof,
-                opening_challenge,
-                Some(rng),
-            )?;
+            result &= Self::check(vk, comms, *point, values, &proof, opening_challenge, rng)?;
```
Comment (Pratyush) [inline src/lib.rs:347]: Same here

File: src/lib.rs [hunk 9]
```diff
@@ -423,16 +458,173 @@ pub trait PolynomialCommitment<F: Field>: Sized {
 
         Ok(true)
     }
+
+    /// open but with individual challenges
+    /// By default, we downgrade them to only use the first individual opening challenges
```
Comment (Pratyush) [inline src/lib.rs:463]: What does this mean? Does it mean that there's one challenge per polynomial? If so, what does it mean to `downgrade them to only use the first individual opening challenges`?

Comment (Pratyush) [reply src/lib.rs:463]: Ah I understand now.

Maybe we should rephrase this to say that `By default, this method invokes `Self::open` with `opening_challenge = opening_challenges[0]?`.

File: src/lib.rs [hunk 10]
```diff
@@ -423,16 +458,173 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
+            point,
+            values,
+            proof,
+            opening_challenges(0),
+            rng,
+        )
+    }
+
+    /// batch_check but with individual challenges
+    /// By default, we downgrade them to only use the first individual opening challenges
```
Comment (Pratyush) [inline src/lib.rs:514]: Same.

File: src/lib.rs [hunk 11]
```diff
@@ -423,16 +458,173 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
+            query_set,
+            evaluations,
+            proof,
+            opening_challenges(0),
+            rng,
+        )
+    }
+
+    /// open_combinations but with individual challenges
+    /// By default, we downgrade them to only use the first individual opening challenges
```
Comment (Pratyush) [inline src/lib.rs:539]: Same.

File: src/lib.rs [hunk 12]
```diff
@@ -423,16 +458,173 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
+            commitments,
+            query_set,
+            opening_challenges(0),
+            rands,
+            rng,
+        )
+    }
+
+    /// check_combinations but with individual challenges
+    /// By default, we downgrade them to only use the first individual opening challenges
```
Comment (Pratyush) [inline src/lib.rs:567]: Same.

File: src/lib.rs [hunk 13]
```diff
@@ -423,16 +458,173 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
+            query_set,
+            evaluations,
+            proof,
+            opening_challenges(0),
+            rng,
+        )
+    }
+
+    /// batch_open but with individual challenges
+    /// By default, we downgrade them to only use the first individual opening challenges
```
Comment (Pratyush) [inline src/lib.rs:594]: Same. (sorry for so many "Same.`" comments!)

File: src/marlin_pc/data_structures.rs [hunk 14]
```diff
@@ -143,6 +146,64 @@ impl<E: PairingEngine> ToBytes for VerifierKey<E> {
...
+    /// Information required to enforce degree bounds. Each pair
+    /// is of the form `(degree_bound, shifting_advice)`.
+    /// This is `None` if `self` does not support enforcing any degree bounds.
+    pub prepared_degree_bounds_and_shift_powers: Option<Vec<(usize, Vec<E::G1Affine>)>>,
+    /// The maximum degree supported by the `UniversalParams` `self` was derived
+    /// from.
+    pub max_degree: usize,
+    /// The maximum degree supported by the trimmed parameters that `self` is
+    /// a part of.
+    pub supported_degree: usize,
```
Comment (Pratyush) [inline src/marlin_pc/data_structures.rs:164]: Invoking `Default::default()` here will set the fields to be empty/zero. Are you sure that's the intended behaviour?

File: src/marlin_pc/data_structures.rs [hunk 15]
```diff
@@ -195,6 +256,35 @@ impl<E: PairingEngine> PCCommitment for Commitment<E> {
...
+    Default(bound = ""),
+    Hash(bound = ""),
+    Clone(bound = ""),
+    Debug(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = "")
+)]
+pub struct PreparedCommitment<E: PairingEngine> {
+    pub(crate) prepared_comm: kzg10::PreparedCommitment<E>,
+    pub(crate) shifted_comm: Option<kzg10::Commitment<E>>,
```
Comment (Pratyush) [inline src/marlin_pc/data_structures.rs:271]: The `Defaul` impl will set this to `None`.

File: src/marlin_pc/mod.rs [hunk 16]
```diff
@@ -363,34 +534,36 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr> for MarlinKZG10<E> {
...
-            let challenge_j = opening_challenge.pow([2 * j as u64]);
+            let challenge_j = opening_challenges(opening_challenge_counter);
+            opening_challenge_counter += 1;
 
             assert_eq!(degree_bound.is_some(), rand.shifted_rand.is_some());
 
-            p += (challenge_j, polynomial.polynomial());
-            r += (challenge_j, &rand.rand);
+            p += (challenge_j.clone(), polynomial.polynomial());
+            r += (challenge_j.clone(), &rand.rand);
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:543]: You should be able to replace these `clone()`s with dereferences:
```suggestion
            p += (*challenge_j, polynomial.polynomial());
            r += (*challenge_j, &rand.rand);
```

File: src/marlin_pc/mod.rs [hunk 17]
```diff
@@ -363,34 +534,36 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr> for MarlinKZG10<E> {
...
 
             if let Some(degree_bound) = degree_bound {
                 enforce_degree_bound = true;
                 let shifted_rand = rand.shifted_rand.as_ref().unwrap();
                 let (witness, shifted_rand_witness) = kzg10::KZG10::compute_witness_polynomial(
                     polynomial.polynomial(),
-                    point,
+                    point.clone(),
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:550]:
```suggestion
                    *point,
```

File: src/marlin_pc/mod.rs [hunk 18]
```diff
@@ -363,34 +534,36 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr> for MarlinKZG10<E> {
...
-                let challenge_j_1 = challenge_j * &opening_challenge;
+                let challenge_j_1 = opening_challenges(opening_challenge_counter);
+                opening_challenge_counter += 1;
 
                 let shifted_witness = shift_polynomial(ck, &witness, degree_bound);
 
-                shifted_w += (challenge_j_1, &shifted_witness);
-                shifted_r += (challenge_j_1, shifted_rand);
+                shifted_w += (challenge_j_1.clone(), &shifted_witness);
+                shifted_r += (challenge_j_1.clone(), shifted_rand);
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:559]:
```suggestion
                shifted_w += (*challenge_j_1, &shifted_witness);
                shifted_r += (&challenge_j_1, shifted_rand);
```

File: src/marlin_pc/mod.rs [hunk 19]
```diff
@@ -363,34 +534,36 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr> for MarlinKZG10<E> {
...
                 if let Some(shifted_rand_witness) = shifted_rand_witness {
-                    shifted_r_witness += (challenge_j_1, &shifted_rand_witness);
+                    shifted_r_witness += (challenge_j_1.clone(), &shifted_rand_witness);
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:561]:
```suggestion
                    shifted_r_witness += (*challenge_j_1, &shifted_rand_witness);
```

File: src/marlin_pc/mod.rs [hunk 20]
```diff
@@ -363,34 +534,36 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr> for MarlinKZG10<E> {
...
                 }
             }
         }
         let proof_time = start_timer!(|| "Creating proof for unshifted polynomials");
-        let proof = kzg10::KZG10::open(&ck.powers(), &p, point, &r)?;
+        let proof = kzg10::KZG10::open(&ck.powers(), &p, point.clone(), &r)?;
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:566]: Does `point` need to be cloned here?

```suggestion
        let proof = kzg10::KZG10::open(&ck.powers(), &p, point, &r)?;
```

File: src/marlin_pc/mod.rs [hunk 21]
```diff
@@ -573,22 +756,22 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr> for MarlinKZG10<E> {
...
                 }
 
                 // Some(_) > None, always.
                 hiding_bound = core::cmp::max(hiding_bound, cur_poly.hiding_bound());
-                poly += (*coeff, cur_poly.polynomial());
-                randomness += (*coeff, cur_rand);
-                coeffs_and_comms.push((*coeff, cur_comm.commitment()));
+                poly += (coeff.clone(), cur_poly.polynomial());
+                randomness += (coeff.clone(), cur_rand);
+                coeffs_and_comms.push((coeff.clone(), cur_comm.commitment()));
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:766]:
```suggestion
                poly += (*coeff, cur_poly.polynomial());
                randomness += (*coeff, cur_rand);
                coeffs_and_comms.push((*coeff, cur_comm.commitment()));
```

File: src/lib.rs (hunk 9 above)
Comment (Pratyush) [reply src/lib.rs:463]: Actually, I think we should reverse the default implementation here, if possible: `open` should call `open_individual_opening_challenges` with `opening_challenges = |pow| opening_challenge.pow(&[pow]);`, i.e. the same impl as in `MarlinKZG`.

File: src/kzg10/data_structures.rs (hunk 1 above)
Comment (npwardberkeley) [reply src/kzg10/data_structures.rs:96]: @weikengchen do you know why?

File: src/marlin_pc/data_structures.rs (hunk 14 above)
Comment (npwardberkeley) [reply src/marlin_pc/data_structures.rs:164]: @weikengchen Do you know about this as well?

File: src/marlin_pc/mod.rs (hunk 16 above)
Comment (npwardberkeley) [reply src/marlin_pc/mod.rs:543]: the compiler complains for me...

File: src/kzg10/data_structures.rs (hunk 1 above)
Comment (weikengchen) [reply src/kzg10/data_structures.rs:96]: Yes. We should change to E::G1Prepared (without Vec).

The original intention here is to match the constraint world where the prepared key may only support very few bits---but in the end, the constraint world opts to support all the bits, due to the implementation for linear combination.

Note: after changing this, we also need to change the AllocVar for the PreparedVerifierKeyVar in the constraint world, which is in a separate PR.

File: src/lib.rs (hunk 3 above)
Comment (weikengchen) [reply src/lib.rs:96]: I agree. I think these two TODOs are ready to drop.

File: src/lib.rs (hunk 4 above)
Comment (weikengchen) [reply src/lib.rs:103]: I will also update the documentation later today to reflect this change.

The extra String is the "meaningful label" for the point (F), like "beta", which is necessary for the constraint world to be data-independent.

File: src/lib.rs (hunk 7 above)
Comment (weikengchen) [reply src/lib.rs:292]: I agree. This is a mixup that conflicts with the upstream change to make the rng optional.

File: src/marlin_pc/data_structures.rs (hunk 14 above)
Comment (weikengchen) [reply src/marlin_pc/data_structures.rs:164]: Let me double-check---especially if we use the prepared verifier key's default somewhere.

File: src/lib.rs [hunk 22]
```diff
@@ -44,6 +31,9 @@ use std::{
 pub mod data_structures;
 pub use data_structures::*;
 
+mod pc_constraints;
+pub use pc_constraints::*;
+
```
Comment (Pratyush) [inline src/lib.rs:36]: I think you didn't intend for this?

File: src/marlin_pc/mod.rs [hunk 23]
```diff
@@ -13,6 +13,9 @@ use rand_core::RngCore;
 mod data_structures;
 pub use data_structures::*;
 
+mod constraints;
+pub use constraints::*;
+
```
Comment (Pratyush) [inline src/marlin_pc/mod.rs:18]: Same, I think this is for a later PR?

File: src/lib.rs (hunk 22 above)
Comment (npwardberkeley) [reply src/lib.rs]: Oops yes, thanks

File: src/kzg10/data_structures.rs (hunk 1 above)
Comment (weikengchen) [reply src/kzg10/data_structures.rs:96]: Ok, it is not Vec<E::G1Prepared> because it is just the power series of `g` to help the computation, and it is not used directly as an element for pairing.

File: src/lib.rs (hunk 4 above)
Comment (weikengchen) [reply src/lib.rs]: Updated the documentation. Will be pushed in a PR soon.

File: src/lib.rs [hunk 24]
```diff
@@ -423,16 +458,173 @@ pub trait PolynomialCommitment<F: Field>: Sized {
...
+            commitments,
+            point,
+            opening_challenges(0),
+            rands,
+            rng,
+        )
+    }
+
+    /// check but with individual challenges
+    /// By default, we downgrade them to only use the first individual opening challenges
```
[the remark this answers is not in the document]
Comment (weikengchen) [reply src/lib.rs:474]: Fixed in the incoming commit.

File: src/lib.rs (hunk 9 above)
Comment (weikengchen) [reply src/lib.rs:448]: Fixed in the incoming commit.

File: src/lib.rs (hunk 6 above)
Comment (weikengchen) [reply src/lib.rs]: I removed the `Default` and likely I will make sure Marlin/PCD did not use it.

File: src/marlin_pc/data_structures.rs (hunk 14 above)
Comment (weikengchen) [reply src/marlin_pc/data_structures.rs:164]: Remove the Default trait and the derivation.

File: src/marlin_pc/mod.rs (hunk 16 above)
Comment (weikengchen) [reply src/marlin_pc/mod.rs:543]: Omit the `.clone()` seems to work.

File: src/marlin_pc/mod.rs (hunk 17 above)
Comment (weikengchen) [reply src/marlin_pc/mod.rs:550]: Removed `.clone()`.

(These are mostly due to Clion's automatic suggestion, though the compiler is happy without it).

File: src/marlin_pc/mod.rs (hunk 19 above)
Comment (weikengchen) [reply src/marlin_pc/mod.rs:561]: Removed `.clone()`

**Comment by weikengchen** [comment]
Handled all the comments and managed to make the checks pass.

@Pratyush can you do a final pass?

Also, `no_std` check failed due to `tracking`. Should we do something with it?

File: Cargo.toml [hunk 25]
```diff
@@ -24,9 +24,16 @@ edition = "2018"
...
-ark-std = { git = "https://github.com/arkworks-rs/utils", default-features = false }
 ark-poly = { git = "https://github.com/arkworks-rs/algebra", default-features = false }
-bench-utils = { git = "https://github.com/arkworks-rs/utils" }
+
+ark-std = { git = "https://github.com/arkworks-rs/utils", default-features = false }
+ark-relations = { git = "https://github.com/arkworks-rs/snark", default-features = false }
+ark-r1cs-std = { git = "https://github.com/arkworks-rs/r1cs-std", default-features = false  }
+
+ark-nonnative-field = { git = "ssh://git@github.com/arkworks-rs/nonnative.git" }
+
```
Comment (Pratyush) [inline Cargo.toml:34]: For the time being let's remove this

```suggestion
```

File: src/data_structures.rs [hunk 26]
```diff
@@ -37,6 +37,13 @@ pub trait PCVerifierKey: Clone + core::fmt::Debug {
     fn supported_degree(&self) -> usize;
 }
 
+/// Defines the minimal interface of prepared verifier keys for any polynomial
+/// commitment scheme.
+pub trait PCPreparedVerifierKey<UNPREPARED: PCVerifierKey> {
+    /// prepare
+    fn prepare(vk: &UNPREPARED) -> Self;
```
Comment (Pratyush) [inline src/data_structures.rs:44]:
```suggestion
pub trait PCPreparedVerifierKey<Unprepared: PCVerifierKey> {
    /// prepare
    fn prepare(vk: &Unprepared) -> Self;
```

**Comment by Pratyush** [comment]
The implementation looks mostly good to me, except a couple of questions:

* Instead of having new `PCPreparedVerifierKey` and `PCPreparedCommitment` traits, do you think it would be cleaner to make these associated types on `PCVerifierKey` and `PCCommitment`? The idea would be to add `prepare` method on the latter traits.

* Currently the default impl of `*_individual_challenges` in `PolynomialCommitment` forwards to the existing impls that use the power-series challenges. Do you think it makes more sense to reverse the direction? That is, to make the "power-series challenge" methods forward instead to the "individual challenge" methods? This is what's already happening in `MarlinKZG10` and `SonicKZG10`, right?

**Comment by weikengchen** [comment]
The program with the second point: 

Is it possible that some polynomial commitment schemes would fail or be insecure if the challenges are correlated (e.g., a power series)?

**Comment by Pratyush** [comment]
I don't believe so, but for that case, it's the responsibility of the developer to override it

(All the schemes implemented so far don't care if the challenges correspond to a univariate polynomial or a multivariate one)

**Comment by weikengchen** [comment]
I am on the way to make these changes.

**Comment by weikengchen** [comment]
For the first one, I made some initial attempts. Making the prepared key an associated type of prepared key would be complicated since in the program we also use the prepared key type and its constraint version (PreparedVerifierKeyVar). Using the associated version would lengthen a few places of the code. 

Let me now do something about the second one.

**Comment by weikengchen** [comment]
For the second one, I think the main reason why we did not implement the reverse is because, to do so, we also need to upgrade sonic_pc and ipa_pc.

Should we do that? Or leave it as a todo for the future?

**Comment by weikengchen** [comment]
(I am on the way to make individual challenges for sonic and ipa...)

**Comment by weikengchen** [comment]
OK. I did the changes to make individual opening challenges. Unfortunately, I committed them to the constraint branch. 
(https://github.com/scipr-lab/poly-commit/pull/47/commits/c3cfd3a5cedd1645175966ee97ee990c6f5c831d)

I am in the process of undoing such changes and committing them to this repo...

**Comment by weikengchen** [comment]
Done the second change.

**Comment by weikengchen** [comment]
Still finding how to fix no_std. Any idea?

**Comment by Pratyush** [comment]
Should be fixed by #48; `tracing` is incompatible with the ARM platform I was using to check CI.

File: Cargo.toml [hunk 27]
```diff
@@ -24,9 +24,14 @@ edition = "2018"
 ark-serialize = { git = "https://github.com/arkworks-rs/algebra", default-features = false, features = [ "derive" ] }
 ark-ff = { git = "https://github.com/arkworks-rs/algebra", default-features = false }
 ark-ec = { git = "https://github.com/arkworks-rs/algebra", default-features = false }
-ark-std = { git = "https://github.com/arkworks-rs/utils", default-features = false }
 ark-poly = { git = "https://github.com/arkworks-rs/algebra", default-features = false }
-bench-utils = { git = "https://github.com/arkworks-rs/utils" }
+
+ark-std = { git = "https://github.com/arkworks-rs/utils", default-features = false }
+ark-relations = { git = "https://github.com/arkworks-rs/snark", default-features = false }
+ark-r1cs-std = { git = "https://github.com/arkworks-rs/r1cs-std", default-features = false  }
```
Comment (Pratyush) [inline Cargo.toml:31]:
```suggestion
```

File: Cargo.toml [hunk 28]
```diff
@@ -52,7 +57,7 @@ incremental = true
 debug = true
 
 [features]
-default = ["std", "parallel"]
-std = [ "ark-ff/std", "ark-ec/std", "ark-poly/std", "ark-std/std", "ark-serialize/std" ]
+default = [ "std", "parallel" ]
+std = [ "ark-ff/std", "ark-ec/std", "ark-poly/std", "ark-std/std", "ark-serialize/std", "ark-relations/std", "ark-r1cs-std/std" ]
```
Comment (Pratyush) [inline Cargo.toml:61]:
```suggestion
std = [ "ark-ff/std", "ark-ec/std", "ark-poly/std", "ark-std/std", "ark-serialize/std" ]
```
